### DISCUSSION TITLE
Websocket route inside a class not working

### DISCUSSION THREAD

**Discussion by OmerNassie3** [fastapi/fastapi#7878] [Questions]
I've written a class that handles api requests and websocket connection.
While did it, i've used APIRouter class from fastapi package and decorated all requests methods with the corresponding decorators:
GET = @api.get
websocket: @api.websocket

my problem is the socket cannot be initiated apparently because of the self parameter. I assume that because as a static method it works very good.

As i have more logic to add, i need the method to be within an instance of the class and not as static.

### To Reproduce

Steps to reproduce the behavior with a minimum self-contained file.
```Python
   @api.websocket("/ws")
    async def websocket_endpoint(self=Depends(), websocket: WebSocket = Depends()):
        pass
```

This is the decoration within the class and what it doesn't work.
registering the routes by: 
```Python
def register_routes(api: FastAPI) -> None:
    """Registers REST namespaces with all included routes
    Args:
        api (Api): REST api initiated in global application context
    Returns: None
    """
    api.include_router(CallIntegratorController().api, prefix='/api/call', tags=['Connector'])
```

### Expected behavior

Support websocket endpoint in a class instance just as http

### Environment

- OS: macOS
- FastAPI Version 0.45.0

- Python version 3.7.3

**Comment by dmontagu** [comment]
If you are trying to use a method of a single instance of a class as a route, you need to decorate the method *after* creating the instance. (This is not a bug with FastAPI, it's just how python works.)

So instead of
```python
    @api.websocket("/ws")
    async def websocket_endpoint(self=Depends(), websocket: WebSocket = Depends()):
        pass
```
you need to have
```python
    async def websocket_endpoint(self, websocket: WebSocket = Depends()):
        pass
```
and something more like
```python
def register_routes(api: FastAPI) -> None:
    """Registers REST namespaces with all included routes
    Args:
        api (Api): REST api initiated in global application context
    Returns: None
    """
    controller = CallIntegratorController()
    controller.api.websocket("/ws")(controller.websocket_endpoint)
    api.include_router(controller.api, prefix='/api/call', tags=['Connector'])
```

The above may not be quite right depending on what `controller.api` is and whether `websocket_endpoint` is a method of `CallIntegratorController` or something else; you haven't provided enough context to make it clear. But it should be possible to achieve this.

-----

If you want to create a *new instance* of the class for *each* request (instead of just once during server startup), there is not currently a built-in way to do this. But you can use the `cbv` function from this gist to achieve it: https://gist.github.com/dmontagu/87e9d3d7795b14b63388d4b16054f0ff (see the usage file at the end for details).

I'm planning to release that as part of a fastapi utilities package soon.

**Comment by OmerNassie3** [comment]
Thank you for the quick response.

The ```cbv``` thing works perfectly!

Your suggestion doesn't work but as i implemented the ```cbv``` in the project my method signature is supported as:
```Python
@router.websocket("/ws")
async def websocket_endpoint(self, websocket: WebSocket):
    pass
```

Thanks for the workaround.

**Comment by tiangolo** [comment]
Thanks for the help here @dmontagu ! :clap: :bow: 

Thanks @OmerNassie3  for reporting back and closing the issue :+1:
